File: 0039-combination-sum/0039-combination-sum.java
class Solution {
    List<List<Integer>> res = new ArrayList<>();
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<Integer> list = new ArrayList<>();
        Arrays.sort(candidates);
        dfs(candidates, target, list, 0);
        return res;
    }
    
    private void dfs(int[] candidates, int target, List<Integer> list, int index) {
        if (target == 0) {
            res.add(new ArrayList<Integer>(list));
            return;
        }
        else if (target > 0) {
            for (int i = index; i < candidates.length; i++) {
                list.add(candidates[i]);
                dfs(candidates, target - candidates[i], list, i);
                list.remove(list.size() - 1);
            }
        }
        else {
            return;
        }
    }
}

File: 0039-combination-sum/0039-combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        res = []
        
        def dfs(i, currSum, currList):
            if target == currSum:
                res.append(list(currList))
                return
            if target < currSum:
                return
            for j in range(i, len(candidates)):
                currList.append(candidates[j])
                currSum += candidates[j]
                dfs(j, currSum, currList)
                currList.pop()
                currSum -= candidates[j]
                
        dfs(0, 0, [])
        return res
            